In [ ]:
import pandas as pd
from sklearn.linear_model import LinearRegression

# Datos
datos = {
    'Edad': [20, 25, 40, 50],
    'BMI': [22.5, 27.1, 31.0, 29.5],
    'Seguro': [2000, 3500, 8000, 12000]
}

# DataFrame
df = pd.DataFrame(datos)

# Variables independientes
X = df[['Edad', 'BMI']]

# Variable dependiente
y = df['Seguro']

# Crear modelo
modelo = LinearRegression()

# Entrenar modelo
modelo.fit(X, y)

# Predicción
edad_nueva = 30
bmi_nuevo = 28

prediccion = modelo.predict([[edad_nueva, bmi_nuevo]])

print("Costo estimado del seguro:", prediccion[0])

Costo estimado del seguro: 4980.027605474531


c:\Users\Pablo Perez\anaconda3\envs\DSRNArtificiales\Lib\site-packages\sklearn\utils\validation.py:2749: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


## Actividad cargando un dataset en csv y normalización de datos

In [69]:
# Importando librerías necesarias
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt   # Librería para crear graficas
from sklearn.linear_model import LinearRegression

#Caegando el dataset en un dataframe de pandas
dataset = pd.read_csv('datasets/insurance.csv')

#Eligiendo los valores a evaluar
dataset_edad_cargos = pd.DataFrame(columns = ['age', 'charges', 'bmi'])

#Obtenemos los costos promedio e imc promedio por edad y se crea un dataframe nuevo
for edad in range(18,64):

    cargos = dataset.query('age == '+str(edad)).charges
    cargos_promedio = np.mean(cargos)

    bmi = dataset.query('age == '+str(edad)).bmi
    bmi_promedio = np.mean(bmi)

    nuevo_registro = pd.DataFrame([{'age': edad, 'bmi': bmi_promedio, 'charges': cargos_promedio}])
    dataset_edad_cargos = pd.concat([dataset_edad_cargos, nuevo_registro], ignore_index=True)
    # dataset_edad_cargos = dataset_edad_cargos.concat({'age' : edad, 'charges' : cargos_promedio}, ignore_index = True)


In [71]:
x1 = dataset_edad_cargos.age
x2 = dataset_edad_cargos.bmi
y = dataset_edad_cargos.charges

# Escalamiento de los datos

x1 = x1.values
x2 = x2.values
y = y.values

x1_max = max(x1)
x1_min = min(x1)
x1 = (x1 - x1_min) / (x1_max - x1_min)

x2_max = max(x2)
x2_min = min(x2)
x2 = (x2 - x2_min) / (x2_max - x2_min)

y_max = max(y)
y_min = min(y)
y = (y - y_min) / (y_max - y_min)


In [77]:
#Utilizando el modelo de regresion lineal
valores = {
    "edad": pd.Series( x1),
    "bmi": pd.Series( x2 ),
    "cargos": pd.Series( y)
}
dataframeEntrenamiento = pd.DataFrame( valores )
# Variables independientes
X = dataframeEntrenamiento[['edad', 'bmi'] ]

# Variable dependiente
y = dataframeEntrenamiento[ 'cargos' ]

# Crear modelo
modelo = LinearRegression()

# Entrenar modelo
modelo.fit(X, y)

# Predicción
edad_nueva = ( 31 - x1_min ) / ( x1_max - x1_min )
bmi_nuevo =  ( 31 - x2_min ) / ( x2_max - x2_min )

dfPredic = pd.DataFrame( [ [edad_nueva, bmi_nuevo] ], columns=[ 'edad', 'bmi' ] )
prediccion = modelo.predict(dfPredic)
print(prediccion)
costoInsuranceInv = (prediccion[0] * (y_max - y_min)) + y_min

print("Costo estimado del seguro:", costoInsuranceInv)

[0.38469212]
Costo estimado del seguro: 11383.327232939106


In [78]:
import pandas as pd

df = pd.read_csv('datasets/insurance.csv')

datos_numericos = df[ ['age', 'bmi', 'children', 'charges'] ]

correlacion = datos_numericos.corr('pearson')

print(correlacion)

               age       bmi  children   charges
age       1.000000  0.109272  0.042469  0.299008
bmi       0.109272  1.000000  0.012759  0.198341
children  0.042469  0.012759  1.000000  0.067998
charges   0.299008  0.198341  0.067998  1.000000


In [86]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score



df = pd.read_csv("datasets/insurance.csv")



df_regresion = df[['age', 'bmi', 'smoker', 'charges']].copy()

# Renombrar columnas
df_regresion.columns = ['edad', 'bmi', 'fumador', 'cargos']



df_regresion['fumador'] = df_regresion['fumador'].map({
    'yes': 1,
    'no': 0
})

print("DataFrame:")
print(df_regresion.head())


#Imprimiendo la correlacion de los datos de nuestro dataset
print('------------------------------------')
print('CORRELACION DE LOS DATOS')
print('------------------------------------')
print(df_regresion.corr('pearson'))


X = df_regresion[['edad', 'bmi', 'fumador']].values
y = df_regresion['cargos'].values



edad_max = np.amax(X[:,0])
edad_min = np.amin(X[:,0])

X[:,0] = (X[:,0] - edad_min) / (edad_max - edad_min)



bmi_max = np.amax(X[:,1])
bmi_min = np.amin(X[:,1])

X[:,1] = (X[:,1] - bmi_min) / (bmi_max - bmi_min)



y_max = np.amax(y)
y_min = np.amin(y)

y = (y - y_min) / (y_max - y_min)


df_normalizado = pd.DataFrame({
    'edad_normalizada': X[:,0],
    'bmi_normalizado': X[:,1],
    'fumador': X[:,2],
    'cargos_normalizados': y
})

print("\nDatos normalizados:")
print(df_normalizado.head())



X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)



modelo = LinearRegression()



modelo.fit(X_train, y_train)



print("\nIntercepto:")
print(modelo.intercept_)

print("\nCoeficientes:")
print("Edad:", modelo.coef_[0])
print("BMI:", modelo.coef_[1])
print("Fumador:", modelo.coef_[2])



predicciones = modelo.predict(X_test)


mae = mean_absolute_error(y_test, predicciones)
mse = mean_squared_error(y_test, predicciones)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, predicciones)

print("\n============================")
print("MÉTRICAS DEL MODELO")
print("============================")

print("MAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R²:", r2)



edad_nueva = 30
bmi_nuevo = 28
fumador_nuevo = 1



edad_norm = (edad_nueva - edad_min) / (edad_max - edad_min)

bmi_norm = (bmi_nuevo - bmi_min) / (bmi_max - bmi_min)

nuevo_cliente = [[edad_norm, bmi_norm, fumador_nuevo]]



prediccion_norm = modelo.predict(nuevo_cliente)

# -------- DESNORMALIZAR --------

prediccion_real = (
    prediccion_norm[0] * (y_max - y_min)
) + y_min

print("\n============================")
print("NUEVO CLIENTE")
print("============================")

print("Edad:", edad_nueva)
print("BMI:", bmi_nuevo)
print("Fumador:", fumador_nuevo)

print("\nCosto estimado del seguro:")
print(prediccion_real)

DataFrame:
   edad     bmi  fumador       cargos
0    19  27.900        1  16884.92400
1    18  33.770        0   1725.55230
2    28  33.000        0   4449.46200
3    33  22.705        0  21984.47061
4    32  28.880        0   3866.85520
------------------------------------
CORRELACION DE LOS DATOS
------------------------------------
             edad       bmi   fumador    cargos
edad     1.000000  0.109272 -0.025019  0.299008
bmi      0.109272  1.000000  0.003750  0.198341
fumador -0.025019  0.003750  1.000000  0.787251
cargos   0.299008  0.198341  0.787251  1.000000

Datos normalizados:
   edad_normalizada  bmi_normalizado  fumador  cargos_normalizados
0          0.021739         0.321227      1.0             0.251611
1          0.000000         0.479150      0.0             0.009636
2          0.217391         0.458434      0.0             0.053115
3          0.326087         0.181464      0.0             0.333010
4          0.304348         0.347592      0.0             0.043816

## Actidivad

- Quitar fumador
- Obtener el RMSE
- Agregar nueva variable

In [85]:
# Importando las librerías necesarias para la actividad

import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

# Leyendo los datos del dataset
df = pd.read_csv("datasets/insurance.csv")

# Generando un dataset a partir de los datos seleccionados
df_regresion = df[['age', 'bmi', 'sex', 'charges']].copy()

#Renombrando las columnas
df_regresion.columns = ['edad', 'bmi', 'genero', 'cargos']

#Generando valores binarios para el campo de 'Genero'
df_regresion['genero'] = df_regresion['genero'].map({
    'male': 1,
    'female': 0
})

#Imprimiendo la correlacion de los datos de nuestro dataset
print('------------------------------------')
print('CORRELACION DE LOS DATOS')
print('------------------------------------')
print(df_regresion.corr('pearson'))

#Generando los arreglos de los valores a utilizar
X = df_regresion[['edad', 'bmi', 'genero']].values
y = df_regresion['cargos'].values

#Obteniendo los máximos y mínimos de edad
edad_max = np.amax(X[:,0])
edad_min = np.amin(X[:,0])

#Normalizando los datos de edad
X[:,0] = (X[:,0] - edad_min) / (edad_max - edad_min)

#Obteniendo los máximos y mínimos de BMI
bmi_max = np.amax(X[:,1])
bmi_min = np.amin(X[:,1])

#Normalizando los datos de BMI
X[:,1] = (X[:,1] - bmi_min) / (bmi_max - bmi_min)


#Obteniendo los máximos y mínimos de cargos
y_max = np.amax(y)
y_min = np.amin(y)

#Normalizando los datos de cargos
y = (y - y_min) / (y_max - y_min)

#Generando un dataframe con los datos normalizados
df_normalizado = pd.DataFrame({
    'edad_normalizada': X[:,0],
    'bmi_normalizado': X[:,1],
    'genero': X[:,2],
    'cargos_normalizados': y
})

#Segmentando el dataframe en sets de entrenamiento y de pruebas
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

#Generando el modelo de regresión
modelo = LinearRegression()

#Entrenando el modelo
modelo.fit(X_train, y_train)

#Obteniendo el bias
print("\nIntercepto:")
print(modelo.intercept_)

#Obtenidno los pesos por columna
print("\nCoeficientes:")
print("Edad:", modelo.coef_[0])
print("BMI:", modelo.coef_[1])
print("Genero:", modelo.coef_[2])

#Utilizando los datos de test para generar las predicciones
predicciones = modelo.predict(X_test)

#Obteniendo las métricas del modelo
mae = mean_absolute_error(y_test, predicciones)
mse = mean_squared_error(y_test, predicciones)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, predicciones)

#Imprimiendo las métricas del modelo
print("\n============================")
print("MÉTRICAS DEL MODELO")
print("============================")
print("MAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R²:", r2)


#Generando una prueba para validar el modelo entrenado
edad_nueva = 30
bmi_nuevo = 28
fumador_nuevo = 1

edad_norm = (edad_nueva - edad_min) / (edad_max - edad_min)
bmi_norm = (bmi_nuevo - bmi_min) / (bmi_max - bmi_min)
nuevo_cliente = [[edad_norm, bmi_norm, fumador_nuevo]]

prediccion_norm = modelo.predict(nuevo_cliente)

#desnormalizando la respuesta e imprimirla
prediccion_real = (
    prediccion_norm[0] * (y_max - y_min)
) + y_min

print("\n============================")
print("NUEVO CLIENTE")
print("============================")

print("Edad:", edad_nueva)
print("BMI:", bmi_nuevo)
print("Fumador:", fumador_nuevo)

print("\nCosto estimado del seguro:")
print(prediccion_real)

------------------------------------
CORRELACION DE LOS DATOS
------------------------------------
            edad       bmi    genero    cargos
edad    1.000000  0.109272 -0.020856  0.299008
bmi     0.109272  1.000000  0.046371  0.198341
genero -0.020856  0.046371  1.000000  0.057292
cargos  0.299008  0.198341  0.057292  1.000000

Intercepto:
0.030899672563566433

Coeficientes:
Edad: 0.16469256585453979
BMI: 0.195162009345432
Genero: 0.021673986222882538

MÉTRICAS DEL MODELO
MAE: 0.14601877493033322
MSE: 0.0333883348258728
RMSE: 0.18272475154141762
R²: 0.1559099048700323

NUEVO CLIENTE
Edad: 30
BMI: 28
Fumador: 1

Costo estimado del seguro:
11067.535864115453


## Explicación de las metricas de evaluación
De las tablas de correlación podemos observar que el parámetro de <bold>género</bold> tiene menor correlacion con los datos que la variable de fumador, por lo que podemos inferir que al incluirla dentro de nuestro modelo de datos, nos puede generar menos valores reales al momento de entrenar y poner a prueba el modelo de regresión lineal.

- Tabla de correlación con género
<table>
    <thead>
        <tr>
            <th> </th>
            <th> EDAD </th>
            <th> BMI </th>
            <th> GENERO </th>
            <th> CARGOS </th>
        </tr>
    </thead>
    <tbody>
        <tr><td>edad</td><td>1.000000</td><td>0.109272</td><td>-0.020856</td><td>0.299008</td></tr>
        <tr><td>bmi</td><td> 0.109272</td><td>1.000000</td><td>0.046371</td><td>0.198341</td></tr>
        <tr><td>genero</td><td>-0.020856</td><td>0.046371</td><td>1.000000</td><td>0.057292</td></tr>
        <tr><td>cargos</td><td>0.299008</td><td>0.198341</td><td>0.057292</td><td>1.000000</td></tr>
    </tbody>
</table>

- Tabla de correlación con fumador

<table>
    <thead>
        <tr> <th> </th> <th> EDAD </th> <th> BMI </th> <th> FUMADOR </th> <th> CARGOS </th>
        </tr>
    </thead>
    <tbody>
        <tr><td>edad</td><td>1.000000</td><td>0.109272</td><td>-0.025019</td><td>0.299008</td></tr>
        <tr><td>bmi</td><td>0.109272</td><td>1.000000</td><td>0.003750</td><td>0.198341</td></tr>
        <tr><td>fumador</td><td>-0.025019</td><td>0.003750</td><td>1.000000</td><td>0.787251</td></tr>
        <tr><td>cargos<td>0.299008</td><td>0.198341</td><td>0.787251</td><td>1.000000</td></tr>
    </tbody>
</table>

Esto de igual manera, se puede observar con los valores obtenidos de las métricas por modelo entrenado:

- Métricas con el valor de genero 
<table>
    <tbody>
        <tr><td>MAE</td><td>0.14601877493033322</td></tr>
        <tr><td>MSE</td><td> 0.0333883348258728</td></tr>
        <tr><td>RMSE</td><td>0.18272475154141762</td></tr>
        <tr><td>R²</td><td>0.1559099048700323</td></tr>
    </tbody>
</table>

- Métricas con el valor de fumador 
<table>
    <tbody>
        <tr><td>MAE</td><td>0.06800731719392254</td></tr>
        <tr><td>MSE</td><td> 0.008793436717604542</td></tr>
        <tr><td>RMSE</td><td>0.0937733262586144</td></tr>
        <tr><td>R²</td><td>0.7776932310583375</td></tr>
    </tbody>
</table>

Los valores de las metricas tienden a 0 con el valor de fumador en las métricas MAE, MSE y RMSE, mientras que R² tiende a 1 indicandonos que el porcentaje de predicciones correctas es mayor que con género.

## Obteniendo valores únicamente con EDAD e IMC

In [ ]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

df = pd.read_csv("datasets/insurance.csv")

df_regresion = df[['age', 'bmi','charges']].copy()

# Renombrar columnas
df_regresion.columns = ['edad', 'bmi','cargos']


#Imprimiendo la correlacion de los datos de nuestro dataset
print('------------------------------------')
print('CORRELACION DE LOS DATOS')
print('------------------------------------')
print(df_regresion.corr('pearson'))


X = df_regresion[['edad', 'bmi']].values
y = df_regresion['cargos'].values

edad_max = np.amax(X[:,0])
edad_min = np.amin(X[:,0])

X[:,0] = (X[:,0] - edad_min) / (edad_max - edad_min)



bmi_max = np.amax(X[:,1])
bmi_min = np.amin(X[:,1])

X[:,1] = (X[:,1] - bmi_min) / (bmi_max - bmi_min)



y_max = np.amax(y)
y_min = np.amin(y)

y = (y - y_min) / (y_max - y_min)


df_normalizado = pd.DataFrame({
    'edad_normalizada': X[:,0],
    'bmi_normalizado': X[:,1],
    'cargos_normalizados': y
})

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

modelo = LinearRegression()

modelo.fit(X_train, y_train)

print("\nIntercepto:")
print(modelo.intercept_)

print("\nCoeficientes:")
print("Edad:", modelo.coef_[0])
print("BMI:", modelo.coef_[1])

predicciones = modelo.predict(X_test)

mae = mean_absolute_error(y_test, predicciones)
mse = mean_squared_error(y_test, predicciones)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, predicciones)

print("\n============================")
print("MÉTRICAS DEL MODELO")
print("============================")

print("MAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R²:", r2)

edad_nueva = 30
bmi_nuevo = 28

edad_norm = (edad_nueva - edad_min) / (edad_max - edad_min)
bmi_norm = (bmi_nuevo - bmi_min) / (bmi_max - bmi_min)
nuevo_cliente = [[edad_norm, bmi_norm]]

prediccion_norm = modelo.predict(nuevo_cliente)
# -------- DESNORMALIZAR --------
prediccion_real = (
    prediccion_norm[0] * (y_max - y_min)
) + y_min

print("\n============================")
print("NUEVO CLIENTE")
print("============================")

print("Edad:", edad_nueva)
print("BMI:", bmi_nuevo)

print("\nCosto estimado del seguro:")
print(prediccion_real)

------------------------------------
CORRELACION DE LOS DATOS
------------------------------------
            edad       bmi    cargos
edad    1.000000  0.109272  0.299008
bmi     0.109272  1.000000  0.198341
cargos  0.299008  0.198341  1.000000

Intercepto:
0.04173855780637606

Coeficientes:
Edad: 0.16432367184818558
BMI: 0.1962636060653186

MÉTRICAS DEL MODELO
MAE: 0.14721272767197904
MSE: 0.03348931883860229
RMSE: 0.18300087114164865
R²: 0.1533569292467799

NUEVO CLIENTE
Edad: 30
BMI: 28

Costo estimado del seguro:
10405.058218534416
